### PR TITLE
[MIG] sale_coupon_domain_product_discount: Migration/refactor to 13.0

### PR DESCRIPTION
**Pull request by chienandalu**
We've got this very same requisite, here's my alternative approach to

- #47

The main objective is to have a more simple implementation.

TODO:

- [x] <del>init, uninstall hooks</del> (not needed)
- [x] UI tweaks
- [x] more tests
- [x] Add the feature for strict limits per product
- [x] Change README

@simahawk @francesco-ooops @elvise @GabbasovDinar @isserver1 I'd really appreciate your opinion

TT36131 @Tecnativa

### CODE REVIEW

**Comment by francesco-ooops** [comment]
@GabbasovDinar can you leave a review? thanks

**Review by francesco-ooops — COMMENTED** [review]
@chienandalu I cannot seem to successfully add a promotion program based on product domain to sale order

ping me when you need a functional review :)

**Comment by GabbasovDinar** [comment]
> @chienandalu I cannot seem to successfully add a promotion program based on product domain to sale order
> 
> ping me when you need a functional review :)

Hello @francesco-ooops 

I tested on [runbot](http://oca-sale-promotion-13-0-pr64-fe0588ffda76.runboat.odoo-community.org/web?debug=1#id=5&action=443&model=sale.coupon.program&view_type=form&cids=1&menu_id=250) and it's work for me 

image

File: sale_coupon_domain_product_discount/models/coupon_program.py [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+            and x.discount_apply_on == "specific_products"
+        )
+        programs = super(
+            SaleCouponProgram, self - domain_discount_programs
+        )._filter_not_ordered_reward_programs(order)
+        for program in domain_discount_programs:
+            discount_specific_product_ids = program.with_context(
+                promo_domain_product=program.rule_products_domain
+            ).discount_specific_product_ids
+            if not order.order_line.filtered(
```
Comment (GabbasovDinar) [inline sale_coupon_domain_product_discount/models/coupon_program.py]: not block for me but maybe it's better to use a positive condition?

File: sale_coupon_domain_product_discount/models/coupon_reward.py [hunk 2]
```diff
@@ -1,22 +1,79 @@
...
+        compute="_compute_discount_specific_product_ids",
+        inverse="_inverse_discount_specific_product_ids",
+        search="_search_discount_specific_product_ids",
+        readonly=False,
+        store=False,
+    )
+    # This is where we'll store our real products for regular promos. Indeed, we're
+    # taking the former relation in order to keep former data. This saves us
+    # supercomplex init/uninstall hooks +1 for the ORM
+    discount_specific_product_ids_stored = fields.Many2many(
```
Comment (GabbasovDinar) [inline sale_coupon_domain_product_discount/models/coupon_reward.py]: https://github.com/OCA/odoo-community.org/blob/master/website/Contribution/CONTRIBUTING.rst#38fields

I think no need add additional string after suffix `_ids`. Maybe `stored_discount_specific_product_ids`

**Comment by GabbasovDinar** [comment]
@chienandalu Thank you for the PR, interesting method to solve the task

File: sale_coupon_domain_product_discount/models/coupon_reward.py (hunk 2 above)
Comment (elvise) [reply sale_coupon_domain_product_discount/models/coupon_reward.py]: @chienandalu what do you think ?

**Review by francesco-ooops — APPROVED** [review]
functional review ok

**Review by elvise — APPROVED** [review]
LGTM! :)

File: sale_coupon_domain_product_discount/models/coupon_program.py (hunk 1 above)
Comment (chienandalu) [reply sale_coupon_domain_product_discount/models/coupon_program.py]: Changed :)

File: sale_coupon_domain_product_discount/models/coupon_reward.py (hunk 2 above)
Comment (chienandalu) [reply sale_coupon_domain_product_discount/models/coupon_reward.py]: Perfect :+1: Changed

**Comment by francesco-ooops** [comment]
@chienandalu think it's ready to request merge?

**Comment by chienandalu** [comment]
Sure :)

File: sale_coupon_domain_product_discount/models/coupon_program.py [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+                lambda line: line.product_id in discount_specific_product_ids
+            ):
+                continue
+            programs |= program
```
Comment (rousseldenis) [inline sale_coupon_domain_product_discount/models/coupon_program.py]: I think building a set of ids and returning a browse(ids) is better in terms of performances than recordset union

Comment (chienandalu) [reply sale_coupon_domain_product_discount/models/coupon_program.py]: Thanks for your comment. I fixed it following Odoo's guidelines in their core method patch: https://github.com/odoo/odoo/commit/19a2f84b39779f7bb68804246bfe28d106da812d

Comment (pedrobaeza) [reply sale_coupon_domain_product_discount/models/coupon_program.py]: Is this so performance penalty taking into account the number of matches? In any case, prefetch should be preserved across browses.

Comment (chienandalu) [reply sale_coupon_domain_product_discount/models/coupon_program.py]: I've got the same feeling. Changed anyway :)

**Review by GabbasovDinar — APPROVED** [review]

**Comment by pedrobaeza** [comment]
/ocabot merge nobump

The refactor should be now fw-ported to v14, isn't it?

**Comment by elvise** [comment]
> /ocabot merge nobump
> 
> 
> 
> The refactor should be now fw-ported to v14, isn't it?

@GabbasovDinar ?

**Comment by GabbasovDinar** [comment]
> > /ocabot merge nobump
> > The refactor should be now fw-ported to v14, isn't it?
> 
> @GabbasovDinar ?

yes, I think we can do it. what do you think @francesco-ooops?

**Comment by francesco-ooops** [comment]
> > > /ocabot merge nobump
> > > The refactor should be now fw-ported to v14, isn't it?
> > 
> > @GabbasovDinar ?
> 
> yes, I think we can do it. what do you think @francesco-ooops?

Sure go ahead! You can also close the original PR #47

**Comment by chienandalu** [comment]
> You can also close the original PR https://github.com/OCA/sale-promotion/pull/47

I'd say you can cherry pick my commits on top of your base work ;)

**Pull request merged**